packages/core/tangleold/ratesetter.go

```diff
@@ -238,12 +296,275 @@ loop:
+	initOnce       sync.Once
+}
+
+func NewDeficitRateSetter(tangle *Tangle) *DeficitRateSetter {
+	return &DeficitRateSetter{
+		tangle:         tangle,
+		Events:         newRateSetterEvents(),
+		self:           tangle.Options.Identity.ID(),
+		issuingQueue:   schedulerutils.NewNodeQueue(tangle.Options.Identity.ID()),
+		issueChan:      make(chan *Block),
+		deficitChan:    make(chan float64),
+		excessDeficit:  atomic.NewFloat64(0),
+		ownRate:        atomic.NewFloat64(0),
+		maxRate:        float64(time.Second / tangle.Scheduler.Rate()),
+		shutdownSignal: make(chan struct{}),
+		shutdownOnce:   sync.Once{},
+	}
+}
+
+func (r *DeficitRateSetter) RateSetterEvents() *RateSetterEvents {
+	return r.Events
+}
+
+// Setup sets up the behavior of the component by making it attach to the relevant events of the other components.
+func (r *DeficitRateSetter) Setup() {
+	r.tangle.BlockFactory.Events.BlockConstructed.Attach(event.NewClosure(func(event *BlockConstructedEvent) {
+		if err := r.Issue(event.Block); err != nil {
+			r.Events.Error.Trigger(errors.Errorf("failed to submit to rate setter: %w", err))
+		}
+	}))
+	// update deficit-based rate setting if own deficit is modified
+	r.tangle.Scheduler.Events.OwnDeficitUpdated.Attach(event.NewClosure(func(event *OwnDeficitUpdatedEvent) {
+		r.deficitChan <- event.Deficit
+	}))
+	go r.IssuerLoop()
+}
+
+// Issue submits a block to the local issuing queue.
+func (r *DeficitRateSetter) Issue(block *Block) error {
+	if identity.NewID(block.IssuerPublicKey()) != r.self {
+		return ErrInvalidIssuer
+	}
+	r.initOnce.Do(func() {
+		// initialize mana vectors in cache here when mana vectors are already loaded
+		r.InitializeRate()
+	})
+
+	select {
+	case r.issueChan <- block:
+		return nil
+	case <-r.shutdownSignal:
+		return ErrStopped
+	}
+}
+
+// Shutdown shuts down the RateSetter.
+func (r *DeficitRateSetter) Shutdown() {
+	r.shutdownOnce.Do(func() {
+		close(r.shutdownSignal)
+	})
+}
+
+// Rate returns the rate of the rate setter.
+func (r *DeficitRateSetter) Rate() float64 {
+	return r.ownRate.Load()
+}
+
+// Size returns the size (in blocks) of the issuing queue.
+func (r *DeficitRateSetter) Size() int {
+	return r.issuingQueue.Size()
+}
+
+// Size returns the size (in blocks) of the issuing queue.
+func (r *DeficitRateSetter) Work() int {
+	return r.issuingQueue.Work()
+}
+
+// Estimate estimates the issuing time of new block.
+func (r *DeficitRateSetter) Estimate() time.Duration {
+	// everything is in units of work (bytes) rather than blocks for deficit-based rate setter
+	return time.Duration(lo.Max(0, (float64(r.Work())-r.excessDeficit.Load())/r.ownRate.Load()))
+}
+
+func (r *DeficitRateSetter) IssuerLoop() {
+	var (
+		nextBlockSize  = float64(0.0)
+		lastDeficit    = float64(0.0)
+		lastUpdateTime = float64(time.Now().Nanosecond()) / 1000000000
+	)
+loop:
+	for {
+		select {
+		// if own deficit changes, check if we can issue something
+		case excessDeficit := <-r.deficitChan:
```

Comment:
Create utility methods to handle each case to make the switch/case more readable.